src/logic/shift.rs

```diff
@@ -0,0 +1,911 @@
+                if width_bits != 0 {
+                    x[x.len() - 1].sign_extend_from(width_bits).unwrap();
+                }
+                let diff = x.len() - digits;
+                if digits == 0 {
+                    //subdigit shift
+                    for i in 0..(x.len() - 1) {
+                        x[i] = (x[i] >> bits) | (x[i + 1] << uns);
+                    }
+                    x[x.len() - 1] = (x[x.len() - 1] >> bits) | (Digit::ONES << uns);
+                } else if bits == 0 {
+                    //digit shift
+                    for i in digits..x.len() {
+                        x[i - digits] = x[i];
+                    }
+                    for i in 0..digits {
+                        x[i + diff].set_all();
+                    }
+                } else {
+                    //digit and subdigit shift
+                    for i in digits..(x.len() - 1) {
+                        x[i - digits] = (x[i] >> bits) | (x[i + 1] << uns);
+                    }
+                    x[diff - 1] = (x[x.len() - 1] >> bits) | (Digit::ONES << uns);
+                    for i in 0..digits {
+                        x[i + diff].set_all();
+                    }
+                }
+            }
+        }
+        self.clear_unused_bits();
+        Ok(())
+    }
+
+    /// Arithmetically right-shifts this `ApInt` by the given `shift_amount` bits
+    /// and returns the result.
+    /// 
+    /// # Errors
+    /// 
+    /// - If the given `shift_amount` is invalid for the bit width of this `ApInt`.
+    pub fn into_wrapping_ashr<S>(self, shift_amount: S) -> Result<ApInt>
+        where S: Into<ShiftAmount>
+    {
+        try_forward_bin_mut_impl(self, shift_amount, ApInt::wrapping_ashr_assign)
+    }
+
+    /// Circularly left-rotates this `ApInt` by the given `shift_amount` bits. In other words, the
+    /// bits are shifted like a logical left shift would, except the bits that go outside the bit
+    /// width of the `ApInt` wrap around to the least significant bits.
+    /// 
+    /// # Errors
+    /// 
+    /// - If the given `shift_amount` is invalid for the bit width of this `ApInt`.
+    /// 
+    /// # Performance
+    /// 
+    /// This function is equivalent to the following:
+    /// ```
+    /// use apint::{ApInt, Width};
+    /// let input = ApInt::from([1u64, 2, 3, 4]);
+    /// let shift = 64usize;
+    /// let output = if shift == 0 {
+    ///     input.clone()
+    /// } else {
+    ///     input.clone().into_wrapping_shl(shift).unwrap()
+    ///         | (&input.clone().into_wrapping_lshr(input.width().to_usize() - shift).unwrap())
+    /// };
+    /// assert_eq!(output, input.into_rotate_left(shift).unwrap());
+    /// assert_eq!(output, ApInt::from([2, 3, 4, 1u64]));
+    /// ```
+    /// 
+    /// However, this function avoids allocation and has many optimized branches for different input
+    /// sizes and shifts.
+    pub fn rotate_left_assign<S>(&mut self, shift_amount: S) -> Result<()>
```

Comment:
Could these implementations maybe profit from using https://doc.rust-lang.org/std/primitive.u64.html#method.rotate_left and https://doc.rust-lang.org/std/primitive.slice.html#method.rotate_left under the hood? I guess they could a lot!

Reply:
Funnily enough, I ended up not using any core bitwise rotate functions due to the way this algorithm played out. The single `Digit` case, for example, did not use it because of support for odd width `ApInt`s.
In the future, I will use the slice rotate function. I am keeping the current one for now because it is faster than the standard one for 1024 bit `ApInt`s or less. I will eventually submit a PR to improve the standard library slice rotate function, and once that is approved I will submit a PR to simplify the `apint` function a little.

Reply:
After your work on the standard rotation functions would this be worth a try again?

Reply:
The `Digit` wise rotation code in question currently uses the standard rotation function, and thus will benefit. The rest of the code deals with bitwise stuff.